libuavcan/include/uavcan/time.hpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright (C) 2014 Pavel Kirienko <pavel.kirienko@gmail.com>
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Types and utilities for working with time.
+ */
+/** @file */
+
+#ifndef UAVCAN_TIME_HPP_INCLUDED
+#define UAVCAN_TIME_HPP_INCLUDED
+
+#include "uavcan/uavcan.hpp"
+#include "uavcan/util/math.hpp"
+
+namespace uavcan
+{
+/**
+ * Protected base class for duration values.
+ *
+ * @tparam DURATIONTYPE  The datatype returned when retrieving durations from
+ *                       realizations of this base class. This type must be
+ *                       exactly 8 bytes in size and must support certain concepts
+ *                       implied by DurationBase.
+ */
+template <typename DURATIONTYPE>
+class DurationBase
+{
+    int64_t usec_; /**< Internal storage of the duration value in microseconds. */
+
+protected:
+    /**
+     * Non-virtual destructor.
+     */
+    ~DurationBase() = default;
+
+    DurationBase()
+        : usec_(0)
+    {
+        static_assert(sizeof(DURATIONTYPE) == 8, "DURATIONTYPE must be 64 bits wide.");
+    }
+
+public:
+    static DURATIONTYPE getInfinite()
+    {
+        return fromUSec(std::numeric_limits<int64_t>::max());
+    }
+
+    static DURATIONTYPE fromUSec(int64_t us)
+    {
+        DURATIONTYPE d;
+        d.usec_ = us;
+        return d;
+    }
+
+    static DURATIONTYPE fromMSec(int64_t ms)
+    {
+        return fromUSec(ms * 1000);
+    }
+
+    int64_t toUSec() const
+    {
+        return usec_;
+    }
+
+    int64_t toMSec() const
```

Comment:
I am mildly irked by the incompatibility of this naming convention with metric prefixes. SI and metric prefixes are no laughing matter.

Shall we rename these into `toMillisecond()`, `toMicrosecond()`, and so on?

Reply:
sure! Will change.